### PR TITLE
Add configurable learning-rate scheduling with cosine annealing and linear warmup

### LINKED ISSUE #282: Add configurable learning-rate scheduling with cosine annealing and linear warmup
ARModel.configure_optimizers() currently returns a bare torch.optim.AdamW with a fixed learning rate for the entire training run:

``` python
# neural_lam/models/ar_model.py  L201-205
def configure_optimizers(self):
    opt = torch.optim.AdamW(
        self.parameters(), lr=self.args.lr, betas=(0.9, 0.95)
    )
    return opt
```
This means neural-lam trains with a constant lr=1e-3 for all 200 epochs, regardless of model size or dataset. There is no option to configure any learning rate schedule — no warmup, no decay, nothing.

PyTorch Lightning natively supports returning {"optimizer": ..., "lr_scheduler": ...} from 
configure_optimizers, but this capability is currently unused.

## Why it matters
- Training stability: Graph neural networks with multi-step autoregressive unrolling (where gradients flow through ar_steps forward passes) are highly sensitive to learning rate. A warmup phase prevents early gradient explosions before the model's residual connections stabilize.
- Convergence quality: Cosine annealing has been shown to yield better minima than constant LR in virtually all modern deep learning training. For weather prediction, this directly translates to lower forecast RMSE at longer lead times.
- Scientific reproducibility: Published NWP-ML papers specify exact LR schedules. Without scheduler support, neural-lam users cannot reproduce published results and must resort to manual hacks:
1.GraphCast (Lam et al., 2023): 1000-step linear warmup → cosine decay to 0
2.Pangu-Weather (Bi et al., 2023): cosine annealing with warmup
3.GenCast (Price et al., 2024): linear warmup → cosine decay
4.FourCastNet (Pathak et al., 2022): cosine annealing with warmup
- Zero performance cost: This is a pure configuration change — no additional compute overhead.

### PR DESCRIPTION
**Pull request by abhaygoudannavar**
## Describe your changes
Add configurable learning rate scheduling to neural-lam's training pipeline, supporting cosine annealing with optional linear warmup.

## Changes:

- neural_lam/train_model.py: Add --lr_scheduler (constant/cosine, default constant), --lr_warmup_epochs (default 0), and --lr_min (default 1e-6) CLI arguments.
- neural_lam/models/ar_model.py: Rewrite configure_optimizers()
 to optionally return a CosineAnnealingLR scheduler, with LinearLR warmup composed via SequentialLR. When --lr_scheduler constant (the default), behavior is identical to before — a bare AdamW optimizer is returned. Also log the current learning rate (lr) in training_step for WandB/MLflow diagnostics.
- tests/test_lr_scheduler.py(new): 4 tests — constant returns bare optimizer, cosine returns scheduler dict, cosine with warmup returns SequentialLR, and an end-to-end training test with cosine + warmup asserting no NaN loss.

- Motivation: configure_optimizers() previously returned a bare AdamW with a fixed lr=1e-3 for the entire training run. All major published weather-ML models use learning rate scheduling:
1.GraphCast (Lam et al., 2023): 1000-step linear warmup → cosine decay to 0
2.Pangu-Weather (Bi et al., 2023): cosine annealing with warmup
3.GenCast (Price et al., 2024): linear warmup → cosine decay
4.FourCastNet (Pathak et al., 2022): cosine annealing with warmup

Without scheduler support, users cannot reproduce published results. A warmup phase also prevents early gradient instability during multi-step autoregressive training.

Dependencies: None — uses only built-in PyTorch (torch.optim.lr_scheduler) and PyTorch Lightning functionality.

Issue Link
Closes #282 

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by leifdenby** [comment]
Thanks for taking this on :) This is already being discussed in https://github.com/mllam/neural-lam/pull/107 though :) Please join the discussion on there and see if something is missing. I will close this issue for now, we can re-open if you think it needs to be its own issue.
